config/src/test/scala/com/typesafe/config/impl/ConfigDocumentParserTest.scala

```diff
@@ -0,0 +1,258 @@
+      parseTest(
+          """
+          foo.bar.baz : {
+            qux : "abcdefg"
+            "abc".def."ghi" : 123
+            abc = { foo:bar }
+          }
+          qux = 123.456
+          """)
+
+      // Test that comments can be parsed in maps
+      parseTest(
+         """{
+          foo: bar
+          // This is a comment
+          baz:qux // This is another comment
+         }""")
+
+      // Basic array tests
+      parseTest("[]")
+      parseTest("[foo]")
+
+      // Test trailing comment and whitespace
+      parseTest("[foo,]")
+      parseTest("[foo,]     ")
+
+      // Can parse arrays with all simple types
+      parseTest("""[foo, bar,"qux", 123,123.456, true,false, null, ${a.b}]""")
+      parseTest("""[foo,   bar,"qux"    , 123 ,  123.456, true,false, null,   ${a.b}   ]""")
+
+      // Basic concatenation tests
+      parseTest("[foo bar baz qux]")
+      parseTest("{foo: foo bar baz qux}")
+      parseTest("[abc 123 123.456 null true false [1, 2, 3] {a:b}, 2]")
+
+      // Complex node with all types test
+      parseTest(
+        """{
+          foo: bar baz    qux    ernie
+          // The above was a concatenation
+
+          baz   =   [ abc 123, {a:12
+                                b: {
+                                  c: 13
+                                  d: {
+                                    a: 22
+                                    b: "abcdefg"
+                                    c: [1, 2, 3]
+                                  }
+                                }
+                                },
+                                //The above value is a map containing a map containing a map, all in an array
+                                22,
+                                // The below value is an array contained in another array
+                                [1,2,3]]
+          // This is a map with some nested maps and arrays within it, as well as some concatenations
+          qux {
+            baz: abc 123
+            bar: {
+              baz: abcdefg
+              bar: {
+                a: null
+                b: true
+                c: [true false 123, null, [1, 2, 3]]
+              }
+            }
+          }
+        // Did I cover everything?
+        }""")
+
+      // Can correctly parse a JSON string
+      val origText =
+        """{
+              "foo": "bar",
+              "baz": 123,
+              "qux": true,
+              "array": [
+                {"a": true,
+                 "c": false},
+                12
+              ]
+           }
+      """
+      val node = ConfigDocumentParser.parse(tokenize(origText), ConfigParseOptions.defaults().setSyntax(ConfigSyntax.JSON))
+      assertEquals(origText, node.render())
+  }
+
+  @Test
+  def parseJSONFailures() {
+      // JSON does not support concatenations
+      parseJSONFailuresTest("""{ "foo": 123 456 789 } """, "Expecting close brace } or a comma")
```

Comment:
All of these exceptions yield the same messages as the ones being thrown by the current parser, with the exception of this one. However, I'm thinking this error message is somewhat more informative (the error message yielded by the other parser states that `' '` is an invalid token in JSON)